### ISSUE TITLE
How to use graphql subscription?

### ISSUE THREAD

**Issue by Subhajitdas298** [leangen/graphql-spqr#140]
Can anyone give me a practical project example? Not a test? Please.

**Comment by kaqqao** [comment]
The closest thing to a practical example is the [SPQR Spring Boot Starter sample](https://github.com/leangen/graphql-spqr-samples/blob/master/spring-boot-starter-sample/src/main/java/io/leangen/graphql/samples/service/IssueService.java#L62). It has [a forked version of GraphiQL](https://github.com/leangen/graphql-spqr-spring-boot-starter/blob/master/graphql-spqr-spring-boot-autoconfigure/src/main/resources/graphiql.html#L211) which add an extra panel at the bottom where you can send subscription requests over WebSocket and where new server pushes are printed. But even that is far from a usable thing.

SPQR itself only deals with the GraphQL server part. Once you obtains the `Publisher`, its job is done. But to make a client able to use it, you need to open a WebSocket, have some kind of protocol established with the server (only Apollo client even [defines](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) such a thing), manage the active sockets on the server etc. There's quite a bit of plumbing before it's usable. 

This is all in scope for the Spring Starter project (as it deals with the web, unlike SPQR itself) but is still work in progress. The aim is to have [Apollo graphql-ws protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) supported out of the box. As of this moment, you'd have to implement it yourself, in addition to the WebSocket backend yourself.

**Comment by kaqqao** [comment]
A few kind people have lately been helping me implement better support for this in the Spring Boot Starter project... so if you're patient a bit more, we'll probably have something more realistic soon.

**Comment by dwardrian** [comment]
which WS port does the above use? Thanks

**Comment by kaqqao** [comment]
@dwardrian  If I'm not mistaken, it's the same port configured for the rest of the app (controlled by Spring Boot). You can override it using the usual `server.port` property in `application.properties`.

**Comment by dwardrian** [comment]
If I run a query or mutation in the playground/ GUI module it works fine but running a subscription returns an error: Could not connect to websocket endpoint ws://localhost:<port>/graphql. Please check the endpoint URL is correct

It seems to detect a web socket connection is needed and uses the same port as http as defined in application properties but as if ws is not being exposed

**Comment by zavatar** [comment]
@dwardrian I found I had a wrong config with target-ws-endpoint.

**Comment by aschonbe** [comment]
@dwardrian What browser were you using for the Playground GUI? It seems there might be issues with that on Firefox. I had the same issue on Firefox and it worked fine on Chome & Safari. Was able to reproduce it on multiple machines (Mac & Linux).

**Comment by kicktipp** [comment]
Indeed. It is a problem with Firefox only. See 
https://github.com/leangen/graphql-spqr-spring-boot-starter/issues/54

**Comment by MathieuPaillart** [comment]
I got the same error : 

{
  "error": "Could not connect to websocket endpoint ws://localhost:8080/graphql. Please check if the endpoint url is correct."
}

I am using chrome

**Comment by pr17m** [comment]
> If I run a query or mutation in the playground/ GUI module it works fine but running a subscription returns an error: Could not connect to websocket endpoint ws://localhost:/graphql. Please check the endpoint URL is correct
> 
> It seems to detect a web socket connection is needed and uses the same port as http as defined in application properties but as if ws is not being exposed

You have to include  "<artifactId>spring-boot-starter-websocket</artifactId>"  in your dependencies

**Comment by hantsy** [comment]
@kaqqao  I am very confused about the WebSocket subscription implementation in GraphQL SPQR. 

I have used some Apollo WebSocket subscription protocol implementations, such a Vertx GraphQl, Netflix Dgs framework, it worked as expected when sending a text message payload according to  Apollo WebSocket subscription protocol, but it did not work in this SPQR. 

GraphQL SPQR ignores the message payload at all, and only accepts the parameters from the WebSocket connection URL, else it will throw an exception like `query can not be null`(when sending text messaging with query as part of payload). 

I tried to write a test for experiencing the WebSocket subscription protocol, I can not make it work, see: https://github.com/leangen/graphql-spqr-spring-boot-starter/issues/112
